cloud/amazon/ec2_vpc_nat_gateway.py

```diff
@@ -0,0 +1,298 @@
+            return [{'state': result.get('NatGateways')[0].get('State'),
+                    'created': result.get('NatGateways')[0].get('CreateTime')}]
+        # Checking if VPC NAT Gateway matching attributes is present
+        elif all((subnet_id, allocation_id)):
+            result = ec2_client.describe_nat_gateways(Filter=[
+                {
+                    'Name': 'subnet-id',
+                    'Values': [subnet_id]
+                }
+            ])
+            if result.get('NatGateways'):
+                gateway_list = list()
+                for nat_gateway in result.get('NatGateways'):
+                    for nat_gateway_address in nat_gateway.get('NatGatewayAddresses'):
+                        if nat_gateway_address.get('AllocationId') == allocation_id:
+                            gateway_list.append({'state': nat_gateway.get('State'),
+                                                 'created': nat_gateway.get('CreateTime'),
+                                                 'nat_gateway_id': nat_gateway.get('NatGatewayId')})
+                return gateway_list
+        return [{'state': 'absent'}]
+    except botocore.exceptions.ClientError as ce:
+        if "NatGatewayNotFound" in ce.message:
+            return [{'state': 'absent'}]
+        else:
+            module.fail_json(msg=ce.message)
+
+
+def delete_nat_gateway(ec2_client=None, module=None, nat_gateway=None):
+    """ Delete an existing NAT gateway """
+    nat_gateway_address = nat_gateway.get('NatGatewayAddresses')[0]
+    nat_gateway_id = nat_gateway['NatGatewayId']
+    results = dict(changed=True, nat_gateway_id=nat_gateway_id,
+                   public_ip=nat_gateway_address.get('PublicIp'),
+                   private_ip=nat_gateway_address.get('PrivateIp'),
+                   allocation_id=nat_gateway_address.get('AllocationId'))
+    ec2_client.delete_nat_gateway(NatGatewayId=nat_gateway_id)
+    wait = module.params.get('wait')
+    if wait:
+        wait_timeout = time.time() + module.params.get('wait_timeout')
+        while wait_timeout > time.time():
+            nat_gateway_status_list = get_nat_gateway_status_list(ec2_client=ec2_client, module=module)
+            if nat_gateway_status_list[0].get('state') in ('deleted', 'absent'):
+                module.exit_json(**results)
+            else:
+                time.sleep(5)
+        module.fail_json(msg="Waited too long for VPC NAT Gateway to be deleted.")
+    else:
+        module.exit_json(**results)
+
+
+def create_nat_gateway(ec2_client=None, module=None):
+    """ Create the NAT gateway """
+    try:
+        result = ec2_client.create_nat_gateway(SubnetId=module.params.get('subnet_id'),
+                                               AllocationId=module.params.get('allocation_id'))
+        wait = module.params.get('wait')
+        start_time = datetime.datetime.utcnow()
+        if wait:
+            wait_timeout = time.time() + module.params.get('wait_timeout')
```

Comment:
I've had a play and I don't think the waiter is the best solution for the following reasons (aside from poor documentation):
1. The failure message is too basic to be useful, e.g.: _"Waiter NatGatewayAvailable failed: Waiter encountered a terminal failure state"_ At least using a custom wait allows for specific failure messages to be returned to the user.
2. There is no waiter for 'NatGatewayDeleted', so I'd still be forced to implement a custom wait for deletion. I don't think it's safe 'not' to wait in all circumstances as it may be followed by an action that requires it to be absent, e.g. re-use of the EIP.